superset/assets/src/CRUD/CollectionTable.jsx

```diff
@@ -0,0 +1,220 @@
+      [newItem.id]: newItem,
+    });
+  }
+  onFieldsetChange(item) {
+    this.changeCollection({
+      ...this.state.collection,
+      [item.id]: item,
+    });
+  }
+  getLabel(col) {
+    const { columnLabels } = this.props;
+    let label = columnLabels[col] ? columnLabels[col] : col;
+    if (label.startsWith('__')) {
+      label = '';
+    }
+    return label;
+  }
+  changeCollection(collection) {
+    this.setState({ collection });
+    this.props.onChange(Object.keys(collection).map(k => collection[k]));
+  }
+  deleteItem(id) {
+    const newColl = { ...this.state.collection };
+    delete newColl[id];
+    this.changeCollection(newColl);
+  }
+  effectiveTableColumns() {
+    const { tableColumns, allowDeletes, expandFieldset } = this.props;
+    const cols = allowDeletes ? tableColumns.concat(['__actions']) : tableColumns;
+    return expandFieldset ? ['__expand'].concat(cols) : cols;
+  }
+  toggleExpand(id) {
+    this.onCellChange(id, '__expanded', false);
+    this.setState({
+      expandedColumns: {
+        ...this.state.expandedColumns,
+        [id]: !this.state.expandedColumns[id],
+      },
+    });
+  }
+  renderHeaderRow() {
+    const cols = this.effectiveTableColumns();
+    return (
+      <thead>
+        <tr>
+          {this.props.expandFieldset && <th className="tiny-cell" />}
+          {cols.map(col => <th key={col}>{this.getLabel(col)}</th>)}
+          {this.props.allowDeletes && <th className="tiny-cell" />}
+        </tr>
+      </thead>
+    );
+  }
+  renderExpandableSection(item) {
+    const propsGenerator = () => ({ item, onChange: this.onFieldsetChange });
+    return recurseReactClone(this.props.expandFieldset, Fieldset, propsGenerator);
+  }
+  renderCell(record, col) {
+    const renderer = this.props.itemRenderers[col];
+    const val = record[col];
+    const onChange = this.onCellChange.bind(this, record.id, col);
+    return renderer ? renderer(val, onChange, this.getLabel(col)) : val;
+  }
+  renderItem(record) {
+    const { tableColumns, allowDeletes, expandFieldset } = this.props;
+    /* eslint-disable no-underscore-dangle */
+    const isExpanded = !!this.state.expandedColumns[record.id] || record.__expanded;
+    let tds = [];
+    if (expandFieldset) {
+      tds.push(
+        <td key="__expand" className="expand">
+          <i
+            className={`fa fa-caret-${isExpanded ? 'down' : 'right'} text-primary pointer`}
+            onClick={this.toggleExpand.bind(this, record.id)}
+          />
+        </td>);
+    }
+    tds = tds.concat(tableColumns.map(col => (
+      <td key={col}>{this.renderCell(record, col)}</td>
+    )));
+    if (allowDeletes) {
+      tds.push(
+        <td key="__actions">
+          <i
+            className="fa fa-close text-primary pointer"
+            onClick={this.deleteItem.bind(this, record.id)}
+          />
+        </td>);
+    }
+    const trs = [<tr className="row" key={record.id}>{tds}</tr>];
+    if (isExpanded) {
```

Comment:
the cleanest expandable tables I have seen have 
1) separated the row component from the table itself (`Table` + `ExpandableRow`) to decouple and to simplify the components
2) have implemented the expanded row with a trick so that you can encapsulate the expansion in one component. the trick is to wrap every row (previously `tr`) in a `tbody`, which can have 1 or more rows. so you can render something like:

```jsx
// ExpandableRow.jsx

...
return (
  <tbody>
    <tr>{/* tds here */}</tr>
    {this.props.isExpanded && <tr><td colspan={x}>{this.props.expandedContent}</td></tr>
  </tbody>
);
```

Reply:
Will these multiple tbody line up, or do I have to specify hard widths?